File: typescript/src/debug/archive/file-session-archive-storage.ts
/**
 * File-system backed storage for session archives.
 * Uses Node.js fs/promises — no external dependencies.
 */

import { mkdir, readdir, readFile, stat, writeFile } from 'node:fs/promises';
import { join, normalize, resolve } from 'node:path';
import type { ArchivedSessionSummary, SessionArchiveStorage, OutputStreamConsumer } from './session-archive-storage.js';
import { Writable } from 'node:stream';

const EXTENSION = '.archive.gz';

export class FileSessionArchiveStorage implements SessionArchiveStorage {
  private readonly _directory: string;

  constructor(directory: string) {
    this._directory = resolve(directory);
  }

  async storeStreaming(sessionId: string, writer: OutputStreamConsumer): Promise<void> {
    const target = this.archivePath(sessionId);
    const dir = join(target, '..');
    await mkdir(dir, { recursive: true });

    const chunks: Buffer[] = [];
    const writable = new Writable({
      write(chunk: Buffer, _encoding, callback) {
        chunks.push(chunk);
        callback();
      },
    });

    await writer(writable);
    await writeFile(target, Buffer.concat(chunks));
  }

  async list(limit: number, prefix?: string): Promise<readonly ArchivedSessionSummary[]> {
    try {
      await stat(this._directory);
    } catch {
      return [];
    }

    const results: ArchivedSessionSummary[] = [];
    let prefixDirs: string[];
    try {
      prefixDirs = await readdir(this._directory);
    } catch {
      return [];
    }

    for (const prefixDir of prefixDirs) {
      const prefixPath = join(this._directory, prefixDir);
      try {
        const s = await stat(prefixPath);
        if (!s.isDirectory()) continue;
      } catch {
        continue;
      }

      // Skip subdirectories that can't contain matches
      if (prefix && prefixDir.length === 1 && prefixDir[0] !== prefix[0]) continue;

      let files: string[];
      try {
        files = await readdir(prefixPath);
      } catch {
        continue;
      }

      for (const file of files) {
        if (!file.endsWith(EXTENSION)) continue;
        const sessionId = file.slice(0, -EXTENSION.length);
        if (prefix && !sessionId.startsWith(prefix)) continue;

        const filePath = join(prefixPath, file);
        try {
          const fileStat = await stat(filePath);
          const relativePath = join(prefixDir, file);
          results.push({
            sessionId,
            key: relativePath,
            sizeBytes: fileStat.size,
            lastModified: fileStat.mtimeMs,
          });
        } catch {
          // skip unreadable files
        }
      }
    }

    results.sort((a, b) => b.lastModified - a.lastModified);
    return results.slice(0, limit);
  }

  async retrieve(sessionId: string): Promise<Buffer> {
    const path = this.archivePath(sessionId);
    try {
      return await readFile(path);
    } catch {
      throw new Error(`Archive not found for session: ${sessionId}`);
    }
  }

  isAvailable(): boolean {
    return true;
  }

  private archivePath(sessionId: string): string {
    if (!sessionId || !sessionId.trim()) {
      throw new Error(`Invalid session ID: ${sessionId}`);
    }
    const prefix = sessionId[0]!;
    const resolved = normalize(join(this._directory, prefix, sessionId + EXTENSION));
    if (!resolved.startsWith(this._directory)) {
      throw new Error(`Session ID escapes archive directory: ${sessionId}`);
    }
    return resolved;
  }
}

File: java/src/main/java/org/libpetri/debug/archive/FileSessionArchiveStorage.java
package org.libpetri.debug.archive;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * File-system backed storage for session archives.
 *
 * <p>Stores archives at {@code {directory}/{prefix}/{sessionId}.archive.lz4}
 * where prefix is the first character of the session ID (UUID hex digit -> 16 folders).
 */
public class FileSessionArchiveStorage implements SessionArchiveStorage {

    private static final String EXTENSION = ".archive.lz4";
    private final Path directory;

    public FileSessionArchiveStorage(Path directory) {
        this.directory = directory;
    }

    @Override
    public void storeStreaming(String sessionId, OutputStreamConsumer writer) throws IOException {
        var target = archivePath(sessionId);
        Files.createDirectories(target.getParent());
        try (var out = Files.newOutputStream(target)) {
            writer.writeTo(out);
        }
    }

    @Override
    public List<ArchivedSessionSummary> list(int limit) throws IOException {
        return list(limit, null);
    }

    @Override
    public List<ArchivedSessionSummary> list(int limit, String prefix) throws IOException {
        if (!Files.isDirectory(directory)) {
            return List.of();
        }

        var results = new ArrayList<ArchivedSessionSummary>();
        try (DirectoryStream<Path> prefixDirs = Files.newDirectoryStream(directory)) {
            for (Path prefixDir : prefixDirs) {
                if (!Files.isDirectory(prefixDir)) continue;
                // Skip subdirectories that can't contain matches
                if (prefix != null && !prefix.isBlank()) {
                    var dirName = prefixDir.getFileName().toString();
                    if (dirName.length() == 1 && dirName.charAt(0) != prefix.charAt(0)) continue;
                }
                try (DirectoryStream<Path> files = Files.newDirectoryStream(prefixDir, "*" + EXTENSION)) {
                    for (Path file : files) {
                        var fileName = file.getFileName().toString();
                        var sessionId = fileName.substring(0, fileName.length() - EXTENSION.length());
                        if (prefix != null && !prefix.isBlank() && !sessionId.startsWith(prefix)) continue;
                        var attrs = Files.readAttributes(file, java.nio.file.attribute.BasicFileAttributes.class);
                        results.add(new ArchivedSessionSummary(
                                sessionId,
                                directory.relativize(file).toString(),
                                attrs.size(),
                                attrs.lastModifiedTime().toInstant()
                        ));
                    }
                }
            }
        }

        results.sort(Comparator.comparing(ArchivedSessionSummary::lastModified).reversed());
        return results.size() > limit ? results.subList(0, limit) : results;
    }

    @Override
    public InputStream retrieve(String sessionId) throws IOException {
        var path = archivePath(sessionId);
        if (!Files.exists(path)) {
            throw new IOException("Archive not found for session: " + sessionId);
        }
        return Files.newInputStream(path);
    }

    @Override
    public boolean isAvailable() {
        return true;
    }

    private Path archivePath(String sessionId) {
        if (sessionId == null || sessionId.isBlank()) {
            throw new IllegalArgumentException("Invalid session ID: " + sessionId);
        }
        String prefix = sessionId.substring(0, 1);
        var resolved = directory.resolve(prefix).resolve(sessionId + EXTENSION).normalize();
        if (!resolved.startsWith(directory)) {
            throw new IllegalArgumentException("Session ID escapes archive directory: " + sessionId);
        }
        return resolved;
    }
}
